### ISSUE TITLE
umbra mapping failures on win8 x64

### ISSUE THREAD

**Issue by derekbruening** [DynamoRIO/drmemory#1810]
http://build.chromium.org/p/client.drmemory/builders/win-8_x64-drm/builds/2221/steps/Dr.%20Memory%20ctest/logs/stdio

Multiple tests are failing with errors like this:

```
10: ~~Dr.M~~ unhandled application memory @0x000007f727e2079a
```

```
> grep ^ERROR: build_drmemory-rel-64/logs/DrMemory-*/g*
build_drmemory-rel-64/logs/DrMemory-bitfield.exe.3804.000/global.3804.log:ERROR: new app segment [0x000007f000000000, 0x0000080000000000)'s shadow segment [0x0000005c00000000, 0x0000006000000000) conflicts with app segseg [0x0000005000000000, 0x0000006000000000)
build_drmemory-rel-64/logs/DrMemory-fuzz_buffer.exe.2124.000/global.2124.log:ERROR: new app segment [0x000007f000000000, 0x0000080000000000)'s shadow segment [0x0000005c00000000, 0x0000006000000000) conflicts with app segseg [0x0000005000000000, 0x0000006000000000)
build_drmemory-rel-64/logs/DrMemory-fuzz_buffer.exe.4140.000/global.4140.log:ERROR: new app segment [0x0000000000000000, 0x0000001000000000)'s shadow segment [0x0000002000000000, 0x0000002400000000) conflicts with app segseg [0x0000002000000000, 0x0000003000000000)
build_drmemory-rel-64/logs/DrMemory-fuzz_buffer.exe.4716.000/global.4716.log:ERROR: new app segment [0x000007f000000000, 0x0000080000000000)'s shadow segment [0x0000005c00000000, 0x0000006000000000) conflicts with app segseg [0x0000005000000000, 0x0000006000000000)
build_drmemory-rel-64/logs/DrMemory-fuzz_buffer.exe.7092.000/global.7092.log:ERROR: new app segment [0x000007f000000000, 0x0000080000000000)'s shadow segment [0x0000005c00000000, 0x0000006000000000) conflicts with app segseg [0x000000f000000000, 0x0000010000000000)
build_drmemory-rel-64/logs/DrMemory-hello.exe.6400.000/global.6400.log:ERROR: new app segment [0x000007f000000000, 0x0000080000000000)'s reserve segment [0x0000003700000000, 0x0000003800000000) conflicts with app segment [0x0000003000000000, 0x0000004000000000)
```

**Comment by zhaoqin** [comment]
Hmm, why do we have so many strange app segment:
```
[0x0000005000000000, 0x0000006000000000)
[0x0000002000000000, 0x0000003000000000)
[0x000000f000000000, 0x0000010000000000)
[0x0000003000000000, 0x0000004000000000)
```
It definitely breaks the assumption about how application memory is layouted.
Maybe we have to either use page table or use the Umbra's paper approach.

**Comment by zhaoqin** [comment]
The error is caused by the module `locale.nls`, which is located someplace like `0xffe7230000` and breaks the umbra's memory layout assumption.

**Comment by derekbruening** [comment]
I plan to make pattern mode the default for x64 (probably it should have been the default long ago).  We'll leave it that way until the mapping issues are solved and we have full mode ported.

**Comment by derekbruening** [comment]
This has been showing up on Appveyor, which is on Server 2012 R2 (similar to Win 8.1), as an assert which initially we put into #1712:

https://ci.appveyor.com/project/DynamoRIO/drmemory/build/1.0.30/job/du8ilypc06mckg4m

```
free.shadow => ~~Dr.M~~ ASSERT FAILURE (thread 2252): C:\projects\drmemory\drmemory\shadow.c:277: false (fail to create shadow memory mapping)
		Core dump file for application C:\projects\drmemory\build\build_drmemory-dbg-64\tests\free.exe (2108) created at C:\projects\drmemory\build\build_drmemory-dbg-64\logs\dynamorio\free.exe.2108.00000000.ldmp
		~~Dr.M~~ WARNING: application exited with abnormal code 0xffffffff;  ~~Dr.M~~ ASSERT FAILURE (thread 2252):
		  C:\projects\drmemory\drmemory\shadow.c:277: false (fail to create shadow
		  memory mapping)
```

**Comment by derekbruening** [comment]
In addition to locale.nls, there is a region of stacks as well:

```
shadow_table_init
new segment: app [0x00000000'00000000, 0x00000010'00000000), shadow [0x00000020'00000000, 0x00000024'00000000), reserve [0x00000028'00000000, 0x00000029'00000000)
new segment: app [0x00000090'00000000, 0x000000a0'00000000), shadow [0x00000044'00000000, 0x00000048'00000000), reserve [0x00000031'00000000, 0x00000032'00000000)
ERROR: new app segment [0x00000220'00000000, 0x00000230'00000000)'s shadow segment [0x00000028'00000000, 0x0000002c'00000000) conflicts with app segseg [0x00000000'00000000, 0x00000010'00000000)
```

```
0:000> lm
start             end                 module name
00000000`15000000 00000000`155c9000   dynamorio   (private pdb symbols)  c:\derek\drmemory\git\build_x64_dbg\dynamorio\lib64\debug\dynamorio.pdb
00000222`c34a0000 00000222`c354b000   KERNEL32_222c34a0000   (export symbols)       KERNEL32.dll
00000222`c35c0000 00000222`c37dd000   KERNELBASE_222c35c0000   (export symbols)       KERNELBASE.dll
00007ff7`26510000 00007ff7`26d45000   drmemorylib   (private pdb symbols)  c:\derek\drmemory\git\build_x64_dbg\bin64\debug\drmemorylib.pdb
00007ff7`36d60000 00007ff7`36ef1000   dbghelp    (export symbols)       dbghelp.dll
00007ff7`a64d0000 00007ff7`a6507000   hello      (private pdb symbols)  c:\derek\drmemory\git\build_x64_dbg\tests\hello.pdb
00007ff8`312c0000 00007ff8`314dd000   KERNELBASE   (deferred)             
00007ff8`317b0000 00007ff8`318a5000   ucrtbase   (export symbols)       ucrtbase.dll
00007ff8`34b20000 00007ff8`34bcb000   KERNEL32   (deferred)             
00007ff8`34dd0000 00007ff8`34fa1000   ntdll      (export symbols)       C:\Windows\SYSTEM32\ntdll.dll

0x222'C33D0000	C:\Windows\System32\locale.nls	n/a	locale.nls			
```

So we have locale.nls causing problems.
And our private libs ended up in the same place?

The 0x90' ones look like stacks at 95-96 and 94-95.  Not sure what's at 91.

```
BaseAddress:       00000091be200000
RegionSize:        00000000000bd000
State:             00002000  MEM_RESERVE
Type:              00020000  MEM_PRIVATE

BaseAddress:       00000091be2bd000
RegionSize:        0000000000005000
State:             00001000  MEM_COMMIT
Protect:           00000004  PAGE_READWRITE
Type:              00020000  MEM_PRIVATE

BaseAddress:       00000091be2c2000
RegionSize:        000000000013e000
State:             00002000  MEM_RESERVE
Type:              00020000  MEM_PRIVATE

BaseAddress:       00000091be400000
RegionSize:        00000000000f7000
State:             00002000  MEM_RESERVE
Type:              00020000  MEM_PRIVATE

BaseAddress:       00000091be4f7000
RegionSize:        0000000000003000
State:             00001000  MEM_COMMIT
Protect:           00000104  PAGE_READWRITE + PAGE_GUARD
Type:              00020000  MEM_PRIVATE

BaseAddress:       00000091be4fa000
RegionSize:        0000000000006000
State:             00001000  MEM_COMMIT
Protect:           00000004  PAGE_READWRITE
Type:              00020000  MEM_PRIVATE

BaseAddress:       00000091be500000
RegionSize:        00000000000f9000
State:             00002000  MEM_RESERVE
Type:              00020000  MEM_PRIVATE

BaseAddress:       00000091be5f9000
RegionSize:        0000000000003000
State:             00001000  MEM_COMMIT
Protect:           00000104  PAGE_READWRITE + PAGE_GUARD
Type:              00020000  MEM_PRIVATE

BaseAddress:       00000091be5fc000
RegionSize:        0000000000004000
State:             00001000  MEM_COMMIT
Protect:           00000004  PAGE_READWRITE
Type:              00020000  MEM_PRIVATE

BaseAddress:       00000091be600000
RegionSize:        0000019104c20000
State:             00010000  MEM_FREE
Protect:           00000001  PAGE_NOACCESS

0:000> ~*r
rip=00007ff834e78b44 rsp=00000091be4fcaa8 rbp=0000000000000000
rip=00007ff834e76994 rsp=00000091be5ff6b0 rbp=0000000000000000
```

**Comment by derekbruening** [comment]
ASLR moves the stack and locale.nls regions around.  Here are some more:

I ran -verbose 2 a bunch of times:
```
09:27 AM ~/drmemory/git/build_x64_dbg
% bin64/drmemory -verbose 2 -pattern 0 -batch -- tests/hello.exe
```

```
09:27 AM ~/drmemory/git/build_x64_dbg
% grep -B 4 'ERROR: new app' logs/DrM*/g* | grep 'new' | sed "s/\(0x........\)/\1'/"
logs/DrMemory-free.exe.6560.000/global.6560.log-new segment: app [0x00000000'00000000, 0x0000001000000000), shadow [0x0000002000000000, 0x0000002400000000), reserve [0x0000002800000000, 0x0000002900000000)
logs/DrMemory-free.exe.6560.000/global.6560.log-new segment: app [0x00000060'00000000, 0x0000007000000000), shadow [0x0000003800000000, 0x0000003c00000000), reserve [0x0000002e00000000, 0x0000002f00000000)
logs/DrMemory-free.exe.6560.000/global.6560.log:ERROR: new app segment [0x00000230'00000000, 0x0000024000000000)'s shadow segment [0x0000002c00000000, 0x0000003000000000) conflicts with app segseg [0x0000006000000000, 0x0000007000000000)
logs/DrMemory-hello.exe.1628.000/global.1628.log-new segment: app [0x00000000'00000000, 0x0000001000000000), shadow [0x0000002000000000, 0x0000002400000000), reserve [0x0000002800000000, 0x0000002900000000)
logs/DrMemory-hello.exe.1628.000/global.1628.log-new segment: app [0x00000090'00000000, 0x000000a000000000), shadow [0x0000004400000000, 0x0000004800000000), reserve [0x0000003100000000, 0x0000003200000000)
logs/DrMemory-hello.exe.1628.000/global.1628.log:ERROR: new app segment [0x00000220'00000000, 0x0000023000000000)'s shadow segment [0x0000002800000000, 0x0000002c00000000) conflicts with app segseg [0x0000000000000000, 0x0000001000000000)
logs/DrMemory-hello.exe.3440.000/global.3440.log-new segment: app [0x00000000'00000000, 0x0000001000000000), shadow [0x0000002000000000, 0x0000002400000000), reserve [0x0000002800000000, 0x0000002900000000)
logs/DrMemory-hello.exe.3440.000/global.3440.log-new segment: app [0x000000f0'00000000, 0x0000010000000000), shadow [0x0000005c00000000, 0x0000006000000000), reserve [0x0000003700000000, 0x0000003800000000)
logs/DrMemory-hello.exe.3440.000/global.3440.log-new segment: app [0x00000260'00000000, 0x0000027000000000), shadow [0x0000003800000000, 0x0000003c00000000), reserve [0x0000002e00000000, 0x0000002f00000000)
logs/DrMemory-hello.exe.3440.000/global.3440.log:ERROR: new app segment [0x00007ff0'00000000, 0x0000800000000000)'s shadow segment [0x0000005c00000000, 0x0000006000000000) conflicts with app segseg [0x000000f000000000, 0x0000010000000000)
logs/DrMemory-hello.exe.3544.000/global.3544.log-new segment: app [0x00000000'00000000, 0x0000001000000000), shadow [0x0000002000000000, 0x0000002400000000), reserve [0x0000002800000000, 0x0000002900000000)
logs/DrMemory-hello.exe.3544.000/global.3544.log-new segment: app [0x000000c0'00000000, 0x000000d000000000), shadow [0x0000005000000000, 0x0000005400000000), reserve [0x0000003400000000, 0x0000003500000000)
logs/DrMemory-hello.exe.3544.000/global.3544.log-new segment: app [0x000001f0'00000000, 0x0000020000000000), shadow [0x0000005c00000000, 0x0000006000000000), reserve [0x0000003700000000, 0x0000003800000000)
logs/DrMemory-hello.exe.3544.000/global.3544.log:ERROR: new app segment [0x00007ff0'00000000, 0x0000800000000000)'s shadow segment [0x0000005c00000000, 0x0000006000000000) conflicts with app segseg [0x000001f000000000, 0x0000020000000000)
logs/DrMemory-hello.exe.4624.000/global.4624.log-new segment: app [0x00000000'00000000, 0x0000001000000000), shadow [0x0000002000000000, 0x0000002400000000), reserve [0x0000002800000000, 0x0000002900000000)
logs/DrMemory-hello.exe.4624.000/global.4624.log-new segment: app [0x000000b0'00000000, 0x000000c000000000), shadow [0x0000004c00000000, 0x0000005000000000), reserve [0x0000003300000000, 0x0000003400000000)
logs/DrMemory-hello.exe.4624.000/global.4624.log-new segment: app [0x000001f0'00000000, 0x0000020000000000), shadow [0x0000005c00000000, 0x0000006000000000), reserve [0x0000003700000000, 0x0000003800000000)
logs/DrMemory-hello.exe.4624.000/global.4624.log:ERROR: new app segment [0x00007ff0'00000000, 0x0000800000000000)'s shadow segment [0x0000005c00000000, 0x0000006000000000) conflicts with app segseg [0x000001f000000000, 0x0000020000000000)
logs/DrMemory-hello.exe.4916.000/global.4916.log-new segment: app [0x00000000'00000000, 0x0000001000000000), shadow [0x0000002000000000, 0x0000002400000000), reserve [0x0000002800000000, 0x0000002900000000)
logs/DrMemory-hello.exe.4916.000/global.4916.log-new segment: app [0x00000080'00000000, 0x0000009000000000), shadow [0x0000004000000000, 0x0000004400000000), reserve [0x0000003000000000, 0x0000003100000000)
logs/DrMemory-hello.exe.4916.000/global.4916.log-new segment: app [0x000001f0'00000000, 0x0000020000000000), shadow [0x0000005c00000000, 0x0000006000000000), reserve [0x0000003700000000, 0x0000003800000000)
logs/DrMemory-hello.exe.4916.000/global.4916.log:ERROR: new app segment [0x00007ff0'00000000, 0x0000800000000000)'s shadow segment [0x0000005c00000000, 0x0000006000000000) conflicts with app segseg [0x000001f000000000, 0x0000020000000000)
logs/DrMemory-hello.exe.5032.000/global.5032.log-new segment: app [0x00000000'00000000, 0x0000001000000000), shadow [0x0000002000000000, 0x0000002400000000), reserve [0x0000002800000000, 0x0000002900000000)
logs/DrMemory-hello.exe.5032.000/global.5032.log-new segment: app [0x000000c0'00000000, 0x000000d000000000), shadow [0x0000005000000000, 0x0000005400000000), reserve [0x0000003400000000, 0x0000003500000000)
logs/DrMemory-hello.exe.5032.000/global.5032.log-new segment: app [0x000001f0'00000000, 0x0000020000000000), shadow [0x0000005c00000000, 0x0000006000000000), reserve [0x0000003700000000, 0x0000003800000000)
logs/DrMemory-hello.exe.5032.000/global.5032.log:ERROR: new app segment [0x00007ff0'00000000, 0x0000800000000000)'s shadow segment [0x0000005c00000000, 0x0000006000000000) conflicts with app segseg [0x000001f000000000, 0x0000020000000000)
logs/DrMemory-hello.exe.5528.000/global.5528.log-new segment: app [0x00000000'00000000, 0x0000001000000000), shadow [0x0000002000000000, 0x0000002400000000), reserve [0x0000002800000000, 0x0000002900000000)
logs/DrMemory-hello.exe.5528.000/global.5528.log-new segment: app [0x000000b0'00000000, 0x000000c000000000), shadow [0x0000004c00000000, 0x0000005000000000), reserve [0x0000003300000000, 0x0000003400000000)
logs/DrMemory-hello.exe.5528.000/global.5528.log:ERROR: new app segment [0x00000240'00000000, 0x0000025000000000)'s shadow segment [0x0000003000000000, 0x0000003400000000) conflicts with app segseg [0x000000b000000000, 0x000000c000000000)
logs/DrMemory-hello.exe.5848.000/global.5848.log-new segment: app [0x00000000'00000000, 0x0000001000000000), shadow [0x0000002000000000, 0x0000002400000000), reserve [0x0000002800000000, 0x0000002900000000)
logs/DrMemory-hello.exe.5848.000/global.5848.log-new segment: app [0x00000040'00000000, 0x0000005000000000), shadow [0x0000003000000000, 0x0000003400000000), reserve [0x0000002c00000000, 0x0000002d00000000)
logs/DrMemory-hello.exe.5848.000/global.5848.log:ERROR: new app segment [0x00000140'00000000, 0x0000015000000000)'s shadow segment [0x0000003000000000, 0x0000003400000000) conflicts with app segseg [0x0000004000000000, 0x0000005000000000)
logs/DrMemory-hello.exe.6072.000/global.6072.log-new segment: app [0x00000000'00000000, 0x0000001000000000), shadow [0x0000002000000000, 0x0000002400000000), reserve [0x0000002800000000, 0x0000002900000000)
logs/DrMemory-hello.exe.6072.000/global.6072.log-new segment: app [0x00000010'00000000, 0x0000002000000000), shadow [0x0000002400000000, 0x0000002800000000), reserve [0x0000002900000000, 0x0000002a00000000)
logs/DrMemory-hello.exe.6072.000/global.6072.log-new segment: app [0x00000150'00000000, 0x0000016000000000), shadow [0x0000003400000000, 0x0000003800000000), reserve [0x0000002d00000000, 0x0000002e00000000)
logs/DrMemory-hello.exe.6072.000/global.6072.log:ERROR: new app segment [0x00007ff0'00000000, 0x0000800000000000)'s reserve segment [0x0000003700000000, 0x0000003800000000) conflicts with app segment [0x0000015000000000, 0x0000016000000000)
logs/DrMemory-hello.exe.6848.000/global.6848.log-add new app segment for [0x00007fff'fffe0000, 0x00007fffffff0000)
logs/DrMemory-hello.exe.6848.000/global.6848.log-new segment: app [0x00000000'00000000, 0x0000001000000000), shadow [0x0000002000000000, 0x0000002400000000), reserve [0x0000002800000000, 0x0000002900000000)
logs/DrMemory-hello.exe.6848.000/global.6848.log:ERROR: new app segment [0x00000020'00000000, 0x0000003000000000) conflicts with app segseg [0x0000000000000000, 0x0000001000000000)
logs/DrMemory-hello.exe.7960.000/global.7960.log-new segment: app [0x00000000'00000000, 0x0000001000000000), shadow [0x0000002000000000, 0x0000002400000000), reserve [0x0000002800000000, 0x0000002900000000)
logs/DrMemory-hello.exe.7960.000/global.7960.log-new segment: app [0x00000030'00000000, 0x0000004000000000), shadow [0x0000002c00000000, 0x0000003000000000), reserve [0x0000002b00000000, 0x0000002c00000000)
logs/DrMemory-hello.exe.7960.000/global.7960.log:ERROR: new app segment [0x000001f0'00000000, 0x0000020000000000)'s reserve segment [0x0000003700000000, 0x0000003800000000) conflicts with app segment [0x0000003000000000, 0x0000004000000000)
```

**Comment by derekbruening** [comment]
From 10,283 runs:

```
% grep -h -E '^new segment|ERROR: new' logs/DrMemory-*/g* | sed 's/ERROR://' | awk '{print $4 " " $5}' | sed "s/\(0x........\)/\1'/g;s/).*/)/;" | sort | uniq -c
  10283 [0x00000000'00000000, 0x00000010'00000000)
    671 [0x00000010'00000000, 0x00000020'00000000)
    662 [0x00000020'00000000, 0x00000030'00000000)
    644 [0x00000030'00000000, 0x00000040'00000000)
    623 [0x00000040'00000000, 0x00000050'00000000)
    665 [0x00000050'00000000, 0x00000060'00000000)
    629 [0x00000060'00000000, 0x00000070'00000000)
    672 [0x00000070'00000000, 0x00000080'00000000)
    601 [0x00000080'00000000, 0x00000090'00000000)
    632 [0x00000090'00000000, 0x000000a0'00000000)
    615 [0x000000a0'00000000, 0x000000b0'00000000)
    673 [0x000000b0'00000000, 0x000000c0'00000000)
    659 [0x000000c0'00000000, 0x000000d0'00000000)
    629 [0x000000d0'00000000, 0x000000e0'00000000)
    646 [0x000000e0'00000000, 0x000000f0'00000000)
    629 [0x000000f0'00000000, 0x00000100'00000000)
     66 [0x00000110'00000000, 0x00000120'00000000)
     92 [0x00000120'00000000, 0x00000130'00000000)
    100 [0x00000130'00000000, 0x00000140'00000000)
    125 [0x00000140'00000000, 0x00000150'00000000)
    195 [0x00000150'00000000, 0x00000160'00000000)
    223 [0x00000160'00000000, 0x00000170'00000000)
    272 [0x00000170'00000000, 0x00000180'00000000)
    317 [0x00000180'00000000, 0x00000190'00000000)
    335 [0x00000190'00000000, 0x000001a0'00000000)
    377 [0x000001a0'00000000, 0x000001b0'00000000)
    417 [0x000001b0'00000000, 0x000001c0'00000000)
    459 [0x000001c0'00000000, 0x000001d0'00000000)
    461 [0x000001d0'00000000, 0x000001e0'00000000)
    554 [0x000001e0'00000000, 0x000001f0'00000000)
    593 [0x000001f0'00000000, 0x00000200'00000000)
    525 [0x00000210'00000000, 0x00000220'00000000)
    515 [0x00000220'00000000, 0x00000230'00000000)
    501 [0x00000230'00000000, 0x00000240'00000000)
    461 [0x00000240'00000000, 0x00000250'00000000)
    439 [0x00000250'00000000, 0x00000260'00000000)
    389 [0x00000260'00000000, 0x00000270'00000000)
    344 [0x00000270'00000000, 0x00000280'00000000)
    305 [0x00000280'00000000, 0x00000290'00000000)
    256 [0x00000290'00000000, 0x000002a0'00000000)
    223 [0x000002a0'00000000, 0x000002b0'00000000)
    163 [0x000002b0'00000000, 0x000002c0'00000000)
    123 [0x000002c0'00000000, 0x000002d0'00000000)
     87 [0x000002d0'00000000, 0x000002e0'00000000)
     67 [0x000002e0'00000000, 0x000002f0'00000000)
     23 [0x000002f0'00000000, 0x00000300'00000000)
   6513 [0x00007ff0'00000000, 0x00008000'00000000)
```

So it looks like the ranges are:

```
        [0x00000000'00000000, 0x00000100'00000000)
        [0x00000110'00000000, 0x00000300'00000000)
        [0x00007ff0'00000000, 0x00008000'00000000)
```

Of course that could change in future Windows updates, but if there were a
simple formula tweak that would cover those, it would be the easiest
short-term fix for the current Umbra x64.

**Comment by derekbruening** [comment]
It looks like this works:

```
&0x00000fff'ffffffff +0x00000400'00000000)
app:
[0x00000000'00000000, 0x00000300'00000000)
[0x00007ff0'00000000, 0x00008000'00000000)
shadow(app):
[0x00000400'00000000, 0x00000700'00000000)
[0x000013f0'00000000, 0x00001400'00000000)
shadow(shadow(app)):
[0x00000800'00000000, 0x00000b00'00000000)
[0x000007f0'00000000, 0x00000800'00000000)
```

**Comment by derekbruening** [comment]
Need to change SEGMENT_ macros to vars.  Used only at init time, so shouldn't be a perf hit.

On win10x64.  I had to also fix #1968 and #1969 and then I ran 1000x and the
shadow worked every time, where it failed every 2 runs before:

```
% for ((i=0; i<1000; i++)); do echo $i; bin64/drmemory -pattern 0 -batch -- tests/hello.exe; done > OUT 2>&1
% grep -c 'NO ERRORS' OUT
1000
% grep -c ASSERT OUT
0
```
